jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java

```diff
@@ -263,33 +279,61 @@ URL getAuthenticationUrl(String scope) throws MalformedURLException {
-
-      try (Connection connection = Connection.getConnectionFactory().apply(authenticationUrl)) {
-        Request.Builder requestBuilder =
-            Request.builder().setHttpTimeout(JibSystemProperties.getHttpTimeout());
-        if (authorization != null) {
-          requestBuilder.setAuthorization(authorization);
-        }
-        Response response = connection.get(requestBuilder.build());
-        String responseString = Blobs.writeToString(response.getBody());
-
-        AuthenticationResponseTemplate responseJson =
-            JsonTemplateMapper.readJson(responseString, AuthenticationResponseTemplate.class);
-        if (responseJson.getToken() == null) {
-          throw new RegistryAuthenticationFailedException(
-              registryEndpointRequestProperties.getServerUrl(),
-              registryEndpointRequestProperties.getImageName(),
-              "Did not get token in authentication response from " + authenticationUrl);
-        }
-        return Authorizations.withBearerToken(responseJson.getToken());
-      }
+      AuthenticationResponseTemplate responseJson =
+          credential != null && credential.isOAuth2RefreshToken()
+              ? fetchTokenWithOAuth(scope, credential.getPassword())
+              : fetchTokenWithBasicAuth(scope);
 
+      if (responseJson.getToken() == null) {
+        throw new RegistryAuthenticationFailedException(
+            registryEndpointRequestProperties.getServerUrl(),
+            registryEndpointRequestProperties.getImageName(),
+            "Did not get token in authentication response from " + getAuthenticationUrl(scope));
+      }
+      return Authorizations.withBearerToken(responseJson.getToken());
     } catch (IOException ex) {
       throw new RegistryAuthenticationFailedException(
           registryEndpointRequestProperties.getServerUrl(),
           registryEndpointRequestProperties.getImageName(),
           ex);
     }
   }
+
+  private AuthenticationResponseTemplate fetchTokenWithOAuth(String scope, String token)
+      throws IOException {
+    URL authenticationUrl = new URL(realm);
+
+    try (Connection connection = Connection.getConnectionFactory().apply(authenticationUrl)) {
+      BlobHttpContent formBody =
+          new BlobHttpContent(
+              Blobs.from(getAuthenticationFormUrlEncoded(scope, token)),
+              MediaType.FORM_DATA.toString(),
+              null); // FORM_DATA
+      Request.Builder requestBuilder =
+          Request.builder().setHttpTimeout(JibSystemProperties.getHttpTimeout()).setBody(formBody);
+
+      Response response = connection.post(requestBuilder.build());
+      String responseString = Blobs.writeToString(response.getBody());
+
+      return JsonTemplateMapper.readJson(responseString, AuthenticationResponseTemplate.class);
+    }
+  }
+
+  private AuthenticationResponseTemplate fetchTokenWithBasicAuth(String scope) throws IOException {
+    URL authenticationUrl = getAuthenticationUrl(scope);
```

Comment:
Let's just do `new URL(realm + "?" + getBasicAuthRequestParameters(scope))` after removing `getAuthenticationUrl()`.

Reply:
Here we cannot append to realm since we need to use POST request in this method, the grant_type/service/scope/refresh_token must be send using `x-www-form-urlencoded` content-type:
see https://github.com/containerd/containerd/blob/master/remotes/docker/authorizer.go#L189

Reply:
This is for basic auth, not requiring POST. I'll take care of this later.

Reply:
Done in #1524.